rust/main/utils/run-locally/src/main.rs

```diff
@@ -536,13 +474,151 @@ fn main() -> ExitCode {
+            "CHAINS_TEST2_BATCHCONTRACTADDRESS",
+            multicall_address_string.clone(),
+        )
+        .hyp_env("CHAINS_TEST2_MAXBATCHSIZE", "5")
+        .hyp_env("CHAINS_TEST3_CONNECTION_URL", "http://127.0.0.1:8545")
+        .hyp_env(
+            "CHAINS_TEST3_BATCHCONTRACTADDRESS",
+            multicall_address_string,
+        )
+        .hyp_env("CHAINS_TEST3_MAXBATCHSIZE", "5")
+        .hyp_env("METRICSPORT", RELAYER_METRICS_PORT)
+        .hyp_env("DB", relayer_db.to_str().unwrap())
+        .hyp_env("CHAINS_TEST1_SIGNER_KEY", RELAYER_KEYS[0])
+        .hyp_env("CHAINS_TEST2_SIGNER_KEY", RELAYER_KEYS[1])
+        .hyp_env("CHAINS_SEALEVELTEST1_SIGNER_KEY", RELAYER_KEYS[3])
+        .hyp_env("CHAINS_SEALEVELTEST2_SIGNER_KEY", RELAYER_KEYS[4])
+        .hyp_env("RELAYCHAINS", "invalidchain,otherinvalid")
+        .hyp_env("ALLOWLOCALCHECKPOINTSYNCERS", "true")
+        .hyp_env(
+            "GASPAYMENTENFORCEMENT",
+            r#"[{
+                "type": "minimum",
+                "payment": "1"
+            }]"#,
+        )
+        .arg(
+            "chains.test1.customRpcUrls",
+            "http://127.0.0.1:8545,http://127.0.0.1:8545,http://127.0.0.1:8545",
+        )
+        // default is used for TEST3
+        .arg("defaultSigner.key", RELAYER_KEYS[2]);
+    if config.sealevel_enabled {
+        relayer_env.arg(
+            "relayChains",
+            "test1,test2,test3,sealeveltest1,sealeveltest2",
+        )
+    } else {
+        relayer_env.arg("relayChains", "test1,test2,test3")
+    }
+}
+
+fn restart_relayer(config: &Config, state: &mut State, rocks_db_dir: &TempDir) {
+    let (child, _) = state.agents.get_mut("RLY").expect("No relayer agent found");
+
+    child.kill().expect("Failed to kill relayer");
+
+    let relayer_env = create_relayer(config, rocks_db_dir);
+
+    log!("Restarting relayer...");
+    state.push_agent(relayer_env.spawn("RLY", Some(&AGENT_LOGGING_DIR)));
+}
+
+fn message_retrieval_invariants_met() -> eyre::Result<bool> {
+    let log_file_path = AGENT_LOGGING_DIR.join("RLY-output.log");
+    let relayer_logfile = File::open(log_file_path).unwrap();
+
+    log!("Checking message statuses were retrieved from logs...");
+
+    const RETRIEVED_MESSGE_LOG: &str = "Message status retrieved from db";
+    let matched_logs = get_matched_lines(&relayer_logfile, RETRIEVED_MESSGE_LOG);
+
+    let no_metadata_message_ids: HashSet<_> = matched_logs
+        .iter()
+        .filter(|s| s.contains("CouldNotFetchMetadata"))
+        .filter_map(|s| {
+            // string parse for id and remove duplicates by collecting
+            // into a hashset
+            s.rsplit_once("id\u{1b}[0m\u{1b}[2m=\u{1b}[0m\"")
+                .and_then(|(_, s2)| s2.split_once("\""))
+                .map(|(id, _)| id)
+        })
+        .collect();
+
+    let no_metadata_message_count = no_metadata_message_ids.len();
+    if no_metadata_message_count < ZERO_MERKLE_INSERTION_KATHY_MESSAGES as usize {
```

Comment:
it wasn't obvious to me why we expected to have `CouldNotFetchMetadata` messages in this test even after `termination_invariants_met` until I read the code and realized it's because of these messages that we don't insert into the merkle tree. Could you make sure we have a comment somewhere (preferably where `message_retrieval_invariants_met` is called) that describes what's going on in this test and why it works?